### PR TITLE
[ENH] DynaMixForecaster, zero-shot multivariate foundation model

### LINKED ISSUE #10448: [ENH] Add support for DynaMix
**Provider**: [DurstewitzLab](https://github.com/DurstewitzLab)
**Model**: DynaMix
**Task**: Time-Series Forecasting
**GH Stars:** 23
**Downloads:** untracked
**Released:**: Oct 2025
**License:** GPL-3.0 license
**GitHub repo:** [DurstewitzLab/DynaMix-python](https://github.com/DurstewitzLab/DynaMix-python)
**HF Model Card:** [DurstewitzLab/dynamix](https://huggingface.co/DurstewitzLab/dynamix)
**Paper:** [True Zero-Shot Inference of Dynamical Systems Preserving Long-Term Statistics](https://proceedings.neurips.cc/paper_files/paper/2025/file/1419d8554191a65ea4f2d8e1057973e4-Paper-Conference.pdf)
**PyPI:** not available

**Implementation notes:**

- There is no pypi version, therefore code needs to be forked, but checkout the license.
- Should be able to do zero-shot and multivariate forecasting
- model should be cached when loading

### PR DESCRIPTION
**Pull request by yash-sangwan**
#### Reference Issues/PRs

Closes #10448.
Part of the foundation-models umbrella issue #6177.

#### What does this implement/fix? Explain your changes.

This PR adds `DynaMixForecaster`, an sktime interface to **DynaMix**
([DurstewitzLab](https://github.com/DurstewitzLab/DynaMix-python)) - a
mixture-of-experts dynamical-systems foundation model for **zero-shot,
multivariate** time series forecasting that preserves long-term statistics and
runs efficiently on CPU.

The forecaster inherits from `BaseForecaster`, loads pretrained weights from the
Hugging Face hub (`DurstewitzLab/dynamix`), and requires no training - the series
passed to `fit` is used as context, and `predict` forecasts the requested horizon.

Three notable hurdles were solved, each following an established sktime pattern:

1. **GPL-3.0 license gate (TiRex-style).** DynaMix is GPL-3.0 (copyleft),
   incompatible with sktime's permissive BSD-3 license, and **not on PyPI**. No
   DynaMix source is vendored into sktime; the GPL-licensed `dynamix` package is
   installed separately by the user via
   `pip install git+https://github.com/DurstewitzLab/DynaMix-python`. Construction
   is gated behind an explicit `license_accepted=True` flag (otherwise a
   `ValueError` is raised), mirroring `TiRexForecaster`, with a `print_license()`
   classmethod. `_fit` additionally raises a clear `ImportError` with the git
   install command if the package is absent.

2. **`@_multiton` caching layer for Hugging Face weights.** Model loading is
   wrapped in a `_cached_DynaMix` multiton (`sktime.utils.singleton._multiton`),
   keyed on `(model, device)`, so a single in-memory instance is shared across
   forecasters with the same configuration. Combined with the Hugging Face disk
   cache used by `load_hf_model`, weights are downloaded once and never reloaded
   on subsequent `fit`/`predict` calls - matching the Chronos/MOIRAI approach.

3. **Stochasticity handled via a `seed` parameter.** DynaMix draws from the
   global `torch` RNG, so forecasts drift slightly across calls, which broke the
   exact-equality contract test `test_score`. Following the Chronos pattern, a
   `seed` parameter was added; when set, `_predict` seeds the torch RNG before
   forecasting so repeated forecasts are reproducible. The
   `property:randomness="stochastic"` tag is set, and `get_test_params` supplies
   `seed=42` so the contract tests are deterministic.

#### Does your contribution introduce a new dependency? If yes, which one?

No new hard or extras dependency, and no change to `pyproject.toml`. The
GPL-licensed `dynamix` package is declared as a soft dependency via the
`python_dependencies` tag and installed by the user on demand. It is
intentionally **not** added to `pyproject.toml` extras (not on PyPI / GPL).

#### What should a reviewer concentrate their feedback on?

- The GPL handling: license gate + soft-dependency-only approach (no vendoring),
  to confirm it satisfies sktime's BSD-3 licensing constraints.
- The `(T_C, S, N)` context tensor construction and output reshaping in
  `_predict`, against the upstream `dynamix.model.forecaster.DynaMixForecaster`
  API.
- The `seed` / `property:randomness` choice for reproducibility.

#### Did you add any tests for the change?

Yes.
- `sktime/forecasting/tests/test_dynamix.py` runs the soft-dep-gated
  `check_estimator(DynaMixForecaster, raise_exceptions=True)` - **all 668
  contract tests pass** locally with the model installed, covering univariate,
  multivariate, and hierarchical fit/predict.
- A standalone `test_dynamix_license_gate` verifies the license `ValueError`
  without requiring `torch` or `dynamix`.

Note: CI cannot install `dynamix` (not on PyPI), so the heavy `check_estimator`
test is skipped in CI by design - identical to other source-only foundation
models. It has been validated locally against the real weights.

#### Any other comments?

Docs entry added to `docs/source/api_reference/forecasting.rst` under
*Pre-trained and foundation models – domain agnostic*.

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) (`code`, `doc`).
- [x] I've added myself to the estimator's `authors` and `maintainers` tags.
- [x] The PR title starts with [ENH].

##### For new estimators
- [x] I've added the estimator to the API reference in `docs/source/api_reference/forecasting.rst`.
- [x] I've added an illustrative usage example to the docstring, in a pydocstyle-compliant `Examples` section.
- [x] The estimator relies on a soft dependency; I've set the `python_dependencies` tag and ensured dependency isolation (imports inside methods, soft-dep-gated tests).

### CODE REVIEW

File: sktime/forecasting/tests/test_dynamix.py
Comment (geetu040) [inline sktime/forecasting/tests/test_dynamix.py:1]: this is not needed, the `check_estimator` tests already run in the CI

File: sktime/forecasting/dynamix.py [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+    import torch
+else:
+
+    class torch:
+        """Dummy class if torch is unavailable."""
+
+        bfloat16 = None
+
+        class Tensor:
+            """Dummy class if torch is unavailable."""
```
Comment (geetu040) [inline sktime/forecasting/dynamix.py]: would `sktime.utils.dependencies._safe_import` not work here?

File: sktime/forecasting/dynamix.py [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+        "tests:vm": True,
+    }
+
+    def __init__(
+        self,
+        model="dynamix-3d-alrnn-v1.0",
+        device: str = "cpu",
+        preprocessing_method: str = "delay_embedding",
+        standardize: bool = True,
+        seed=None,
```
Comment (geetu040) [inline sktime/forecasting/dynamix.py]: you should use the tags: `property:randomness`, `capability:random_state`
and parameter `random_state`

File: sktime/forecasting/dynamix.py [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+        X : pd.DataFrame, optional (default=None)
+            Exogenous time series (ignored).
+
+        Returns
+        -------
+        y_predict : pd.DataFrame
+            Point forecasts, same columns as the series seen in ``_fit``.
+        """
+        # alias the upstream forecaster to avoid the name collision with this class
+        from dynamix.model.forecaster import DynaMixForecaster as _DynaMixModel
```
Comment (geetu040) [inline sktime/forecasting/dynamix.py]: we need to fork the relevant code, since it's not available on pypi

**Review by geetu040 — CHANGES_REQUESTED** [review]
Nice, the implementation looks good. I have added some review comments. Also you'd need to fork the library code into `sktime.libs`

**Comment by yash-sangwan** [comment]
Thanks for the review, @geetu040!

I've pushed a commit addressing three of the four points:

* Renamed `seed` to `random_state` and set the `capability:random_state` and   `property:randomness: "derandomized"` tags.
* Switched to `_safe_import` instead of the manual try/except + dummy class.
* Deleted `test_dynamix.py` so CI picks up the `check_estimator` contract tests
  automatically.

On vendoring into `sktime.libs`:

This is the one I'd like your guidance on. DynaMix-python is **GPL-3.0**, whereas `sktime` is BSD-3-Clause. Forking GPL code into `sktime.libs` would mean distributing copyleft code as part of an sktime release, which I believe would conflict with sktime's permissive license (the existing vendored forks I checked are permissively licensed - e.g. Lag-Llama is Apache-2.0). There's also no PyPI release, so it can't be pinned as a normal soft dependency either.

To avoid the licensing issue, I've currently kept it as a user-installed soft dependency (`pip install git+...`), gated behind a `license_accepted` flag that defaults to `False` (same pattern as `TiRexForecaster`). The trade-off is that CI can't install it, so the heavy contract tests are skipped there - which is why I validated `check_estimator` locally (668/668 passing).

Happy to go the vendoring route instead if you've cleared the GPL/BSD question, just let me know which direction you'd prefer.

**Comment by yash-sangwan** [comment]
Hii @geetu040 , just a ping on this..

**Comment by geetu040** [comment]
> On vendoring into `sktime.libs`:
> 
> This is the one I'd like your guidance on. DynaMix-python is **GPL-3.0**, whereas `sktime` is BSD-3-Clause. Forking GPL code into `sktime.libs` would mean distributing copyleft code as part of an sktime release, which I believe would conflict with sktime's permissive license (the existing vendored forks I checked are permissively licensed - e.g. Lag-Llama is Apache-2.0). There's also no PyPI release, so it can't be pinned as a normal soft dependency either.

I think it's fine to fork the code, since we make the user aware of the license before they use it.

```
license_accepted : bool, default=False
    Whether the user accepts the GPL-3.0 license terms of DynaMix.
    Must be set to ``True`` to use the model.
```

CC: @fkiraly

**Comment by yash-sangwan** [comment]
Thanks for clarifying, @geetu040!

**Comment by yash-sangwan** [comment]
Hi @geetu040 (cc @fkiraly),

I have successfully vendored the `DynaMix` inference code into `sktime/libs/dynamix`! 

Here is a quick summary of the updates in the latest commit:
* **Vendored & Trimmed:** Copied only the necessary inference files (omitting the training/plotting scripts to keep the footprint small).
* **Strict CI Isolation:** Replaced all heavy bare imports (`torch`, `safetensors`, `huggingface_hub`, etc.) inside the vendored code with `_safe_import` blocks. This completely plugs the `test_softdeps` that was crashing the pipeline earlier.
* **License Preservation:** Added the explicit GPL-3.0 attribution and full license text to the vendored `__init__.py`, keeping the `license_accepted` flag on the estimator.
* **Wired Up:** Updated `DynaMixForecaster` to route through the new `sktime.libs.dynamix` path and removed the old git-install guard.

I have verified locally that the soft dependency checks and all estimator contract tests are passing perfectly.

**Comment by yash-sangwan** [comment]
Made another commit, CI was failing because the `test-est` matrix installs **only** the dependencies an estimator declares in its `python_dependencies` tag (plus sktime's core deps). `DynaMixForecaster` declared `torch`, `huggingface_hub`, and `safetensors`, but **not `statsmodels`**.

The default `preprocessing_method="delay_embedding"` runs, with `tau=None`, through `delay_embedding - estimate_TDM_tau - acf()`, where `acf` is `statsmodels.tsa.stattools.acf`. Since the vendored code guards it with `_safe_import`, a missing `statsmodels` resolves to a harmless mock at **import** time, but calling it at **predict** time raises, so every fit/predict contract test errored. That's why all 15 jobs failed fast and deterministically.

It passed locally only because the `dynamix` pip package pulls `statsmodels` in transitively, masking the gap.

**Review by fkiraly — CHANGES_REQUESTED** [review]
ok, wait, the vendor has a GPL license, correct?

We should avoid moving GPL licensed content into `sktime`.
If we need a fork, I would suggest moving it to an `sktime-copyleft` repository.

Question: why did we need the fork, if there is the `dynamix` package?

Also, we should introduce a "license warning" parameter in the `__init__` that makes the estimator fail at default, similar for other non-permissive licensed foundation models.

**Comment by yash-sangwan** [comment]
I had the same concern about the GPL/BSD conflict.

My original implementation used a soft dependency (`git+https://github.com/DurstewitzLab/DynaMix-python`) together with a `license_accepted` parameter, similar to `TiRexForecaster`. During the review we explored vendoring the code since DynaMix isn't available on PyPI.

I agree we should avoid vendoring GPL-licensed code. I'll revert that change, remove the code from `sktime/libs`, and switch back to the original soft-dependency approach with the `license_accepted` parameter.

**Comment by fkiraly** [comment]
Install from GitHub is also not good since it introduces security vulnerabilities. I think we should create an `sktime-copyleft-libs` package. I will create a repo for this.

**Comment by yash-sangwan** [comment]
Sounds good. I'll wait for the new repository before updating this PR. Would you prefer I revert the vendored code in the meantime, or switch everything over once the new package is ready?

**Comment by yash-sangwan** [comment]
Just following up on this. Happy to help set up the `sktime-copyleft-libs` repository if that would help, just let me know what you'd like me to do.

**Comment by yash-sangwan** [comment]
Hi @fkiraly, Quick update on the copyleft package.
I tried to create `sktime-libs-copyleft` but the name already exists (I get “name already exists” but 404 when trying to view it, mybe its private). To keep moving I created `sktime-copyleft-libs` instead and have already put the DynaMix inference code there.

Repo: https://github.com/sktime/sktime-copyleft-libs

One blocker remains: the package is not on PyPI yet. Without that we cannot properly update this PR.
Could you help with publishing it.
